tests/phpunit/tests/kses.php

```diff
@@ -1353,6 +1354,11 @@ public function data_safecss_filter_attr() {
 				'css'      => 'background-repeat: no-repeat',
 				'expected' => 'background-repeat: no-repeat',
 			),
+			// Convert "&amp;" to "&".
+			array(
+				'css'      => 'background: url("/sites/2/2023/10/image.jpg?width=1024&amp;height=600")',
+				'expected' => 'background: url("/sites/2/2023/10/image.jpg?width=1024&height=600")',
```

Comment:
For image `src` and link `href` attributes the standard is to use the encoded `&amp;` in the attributes. It would be good to check against the spec whether the correct thing to do here would be to always encode the characters from `&` to `&amp;`. @dmsnell Do you know which it should be?

A test for both forms would be good but can wait until the spec is clarified, ie :

```
'css'      => 'background: url("/sites/2/2023/10/image.jpg?width=1024&amp;height=600")',
'css'      => 'background: url("/sites/2/2023/10/image.jpg?width=1024&height=600")',
```

Reply:
thanks for the poke, @peterwilsoncc 

this is difficult to do programmatically because the conversion doesn't indicate if it happened already.

all URLs in URL-containing attributes _should_ have `&` replaced with `&amp;`, but the specification makes allowance for errors just like this, where someone forgets to replace `&` in query args.

the important things to consider are layers of transformation:
 - when decoding HTML attributes, the `&amp;` will be transformed into `&`
 - the resulting decoded URL will still have `&`

so if we naively replace `&` we can end up with `&amp;amp;`. maybe the best thing to do is to decode the attribute and re-encode if we want to do it programmatically.

otherwise, for hard-coded strings we should always prefer `&amp;` as that's how we nominally represent the underlying `&` in the URL itself.

Reply:
we can use the upcoming HTML Text Decoder from #6387 (which I still hope to get in for 6.6) to decode the attribute, and then re-encode it. after decoding, every `&` is guaranteed to be a `&` and not part of a character reference. actually one hope I have with this decoder is that we can simplify many other bits of logic like this, because then we can stop trying to refer to multiple worlds at the same time and only apply logic to native PHP string values (and not HTML-encoded values)

```php
$css = str_replace( array( "\n", "\r", "\t" ), '', $css );
$css = WP_HTML_Decoder::decode_attribute( $css );

$allowed_prototocols = …


```

Reply:
@dmsnell thanks for the feedback, I have implemented the suggestion please check.

Reply:
Thanks @narenin, though the code I referred to isn’t in Core yet. If all goes well in 6.6 it should be available after that

Reply:
@dmsnell Yes, we can take this once it is available in 6.6 or we can take together.

Reply:
@dmsnell Yes, we can take this once it is available in 6.6 or we can take together.

Reply:
Hey @dmsnell, can you (re)review this change please, now that `WP_HTML_Decoder::decode_attribute()` landed in `trunk` via [r58281](https://core.trac.wordpress.org/changeset/58281)?

Reply:
Thanks @hellofromtonya.

This patch looks fine, and we should be clear for using the decoder. Interestingly enough I don't believe that the call to `wp_kses_no_null()` or the `str_replace()` on (some) of the whitespace characters is necessary or helpful. In fact, I know that it would be safer to avoid calling those (because they remove characters, which joins things that were previously separated), but I don't see any reason to delay this patch.

If we want to remove those lines above the call to the decoder we can, but otherwise this is better than it was before and no worse than it was.